### ISSUE TITLE
comm.py communication issues with python3.7

### ISSUE THREAD

**Issue by alberto99** [maccallumlab/meld#72]
We had an upgrade in our cluster for parallelization protocols and i'm running into issues running a successful meld simulation that seem related to the mpi communication. I have gone back and forth between modules for compilation but i'm now using
 1) cuda/10.0.130   3) openmpi/3.1.2   5) cmake/3.15.6      9) swig/3.0.8  
  2) gcc/7.3.0       4) python/3.7      6) doxygen/1.8.3.1   8) netcdf/4.2 

Simulations used to fail pointing to communication issues and contextlib.py. From what i see, python3.7 changed the StopIteration handling in generators. I have been trying to see what changes in comm.py could reflect this. In the end i just edited the contextlib _GeneratorContextManager so that it does not die on a RaiseException:
```
    def __enter__(self):
        # do not keep args and kwds alive unnecessarily
        # they are only needed for recreation, which is not possible anymore
        del self.args, self.kwds, self.func
        try:
            return next(self.gen)
        except StopIteration:
            #raise RuntimeError("generator didn't yield") from None
            return
```
This was a quick and dirty way of getting simulations to run, but it is not an optimal fix. I suspect that there is something going on with the mpi protocol, but i'm not sure how to diagnose it. Simulations with this fix are taking a 50% performance hit.

Any ideas?Thanks

**Comment by jlmaccal** [comment]
Could you post any tracebacks to see what the exact error is?

**Comment by jlmaccal** [comment]
So, I think this is probably coming from the code in `comm.py` that handles timeouts. This is supposed to abort the whole calculation if one or more nodes takes far longer than is expected to complete one of the communication calls. By default, this is 600 seconds or 10 minutes.  If you remember, we would occasionally get this in the past, but the job would just sit there for days, wasting resources.

I suspect that the code that we use for this is not completely compatible with > 3.5 due to the changes to `StopIteration` that you point out. This is is likely to be an issue only in cases where the timeouts are being triggered, which is probably why we haven't seen this before, as we've been using python 3.7 for a long time now without issue.

I think:
- Your fix isn't going to help anything at all. All you've really done is ignore the time out that is normally supposed to kill your simulation because a step was taking too long. This would also explain why things are running slowly once you make this change.
- I suspect that one of the following is true:
    - You are having some sort of network or MPI issues.
    - You have multiple jobs running on the same GPU or a node is somehow overloaded.

Steps to fix:
- You need to sort out what is causing nodes to sometimes run slowly. Some places to look:
    - Check the log fils, paying close attention to the timestamps. You are looking for places were long periods of time (> 5 minutes) go by.
    - Log into every node where the MELD calculation is running and use `top` and `nvidia-smi` to check for rogue processes. I suspect that there will be something running that shouldn't be, or a bunch of jobs all running on the same GPU, etc.
- We need to sort out how to fix the timeout code to give the correct error. The code I borrwed for this, is now a module called `interruptingcow`, so we could try just using that directly. It's been updated since the release of python 3.5, so should be compatible with changes to `StopIteration` inside generators.

TL;DR I think something is screwed up with your cluster/calculation. What you're seeing is symptom, not the problem. And your fix doesn't help.

**Comment by jlmaccal** [comment]
It would be great if you could try the `comm.py` from #73 to see if that helps.

**Comment by jlmaccal** [comment]
OK, so some of that might have been total nonsense. It really does seem like something was broken, but I don't see any documentation of what changed. It's also possible we were running with an older python version and so didn't see the error.

Try the fix in #74 to see if it fixes things. It's just a one liner.

**Comment by frank-stonybrook** [comment]
We had this issue long time ago. I reported to @jlmaccal a year ago but didn't get attention at that moment. I am glad there is fix right now! Let me know if the fix is working or not.

**Comment by alberto99** [comment]
Thanks Justin, 
the fix i made was not a fix at all, just wanted to see where the issue was. The whole episode thrat triggered this was a change of openmpi protocols from mpix_v1 to only allow mpix_v2 and mpix_v3 protocols in a slurm queueing system. {which i do not understand}. 
So depending on the python i was using and the mpi4py used, a simple
from mpi4py import MPI 
triggered failure (e.g. during system setup). When finding a good combo of mpi4py with our environment and successfully compiling openmm/meld setup runs fine and the error i pointed above happens. With tracebags similar to this:
 File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/meld/comm.py", line 589, in timeout_context_manager
   next(t)
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
 File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/bin/launch_remd", line 4, in <module>
   __import__('pkg_resources').run_script('Meld==0.4.14', 'launch_remd')
 File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/pkg_resources/__init__.py", line 666, in run_script
   self.require(requires)[0].run_script(script_name, ns)
 File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/pkg_resources/__init__.py", line 1462, in run_script
   exec(code, namespace, namespace)
 File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/EGG-INFO/scripts/launch_remd", line 27, in <module>
   main()
 File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/EGG-INFO/scripts/launch_remd", line 23, in main
   launch.launch(console, args.debug, args.console_log)
 File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/meld/remd/launch.py", line 69, in launch
   communicator.broadcast_logger_address_to_followers(logger_address)
 File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/meld/util.py", line 39, in wrapper
   res = func(*args, **kwds)
 File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/meld/comm.py", line 131, in broadcast_logger_address_to_followers
   self._mpi_comm.bcast(address, root=0)
 File "/apps/python/3.7/lib/python3.7/contextlib.py", line 119, in __exit__
   next(self.gen)
RuntimeError: generator raised StopIteration


i had tried to change comm.py into a try: yield next(t) except:   kind of pipeline but was failing there -- that's when i did the hack in contextlib just to see that  messages do get send around from meld leader to followers.

But, i think the bigger issue is with the mpi protocols. I been workin with our system adminstrators who now think is an issue of MPI_THREAD_MULTIPLE vs MPI_THREAD_SINGLE.

We have been at this for a week and i will keep you posted of what we find.

**Comment by frank-stonybrook** [comment]
> Thanks Justin,
> the fix i made was not a fix at all, just wanted to see where the issue was. The whole episode thrat triggered this was a change of openmpi protocols from mpix_v1 to only allow mpix_v2 and mpix_v3 protocols in a slurm queueing system. {which i do not understand}.
> So depending on the python i was using and the mpi4py used, a simple
> from mpi4py import MPI
> triggered failure (e.g. during system setup). When finding a good combo of mpi4py with our environment and successfully compiling openmm/meld setup runs fine and the error i pointed above happens. With tracebags similar to this:
> File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/meld/comm.py", line 589, in timeout_context_manager
> next(t)
> StopIteration
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
> File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/bin/launch_remd", line 4, in 
> **import**('pkg_resources').run_script('Meld==0.4.14', 'launch_remd')
> File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/pkg_resources/**init**.py", line 666, in run_script
> self.require(requires)[0].run_script(script_name, ns)
> File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/pkg_resources/**init**.py", line 1462, in run_script
> exec(code, namespace, namespace)
> File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/EGG-INFO/scripts/launch_remd", line 27, in 
> main()
> File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/EGG-INFO/scripts/launch_remd", line 23, in main
> launch.launch(console, args.debug, args.console_log)
> File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/meld/remd/launch.py", line 69, in launch
> communicator.broadcast_logger_address_to_followers(logger_address)
> File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/meld/util.py", line 39, in wrapper
> res = func(*args, **kwds)
> File "/home/alberto.perezant/Source/VirtualEnvironments/MeldCuda_py3.7/lib/python3.7/site-packages/Meld-0.4.14-py3.7.egg/meld/comm.py", line 131, in broadcast_logger_address_to_followers
> self._mpi_comm.bcast(address, root=0)
> File "/apps/python/3.7/lib/python3.7/contextlib.py", line 119, in **exit**
> next(self.gen)
> RuntimeError: generator raised StopIteration
> 
> i had tried to change comm.py into a try: yield next(t) except: kind of pipeline but was failing there -- that's when i did the hack in contextlib just to see that messages do get send around from meld leader to followers.
> 
> But, i think the bigger issue is with the mpi protocols. I been workin with our system adminstrators who now think is an issue of MPI_THREAD_MULTIPLE vs MPI_THREAD_SINGLE.
> 
> We have been at this for a week and i will keep you posted of what we find.

 I believe that we also encountered the same problem with MVAPICH2 MPI and Intel MPI on Frontera. As a dirty trick, we just comment out the timeout context in every MPI calls in the comm.py script.

**Comment by alberto99** [comment]
That seems dangerous for the reasons that Justin outlined before. I would not want to do any production run with something containing tricks. From my point of view i wanted to show that the problem was in the precompiled mpi communication compatibility, not in meld itself. If your jobs are not hanging, you are probably fine. In my case, as soon as i go from trying on 2 gpu to trying on 30 gpus communications fail and jobs hang.

What was the issue you were seeing? Was it the same  generator raised StopIteration or soemthing else?

**Comment by frank-stonybrook** [comment]
> What was the issue you were seeing? Was it the same generator raised StopIteration or soemthing else?

Yes, exactly the same issue you had saw.

> That seems dangerous for the reasons that Justin outlined before. I would not want to do any production run with something containing tricks.

Sorry, I didn't see the dangerousness of commenting out the timeout context. If I understood correctly, the timeout is a useful debugging mechanism which provide information of problematic follower runs in the REMD and kill it after the default timeout time. In the worst case, the problematic run doesn't get killed, the `exchange_states_for_energy_calc ` function call in the leader.py will still block it and the simulation will keep hanging there. Plus, this mechanism doesn't use in older MELD version.  Besides, we validated it against previous MELD runs and the results are reproducible.

**Comment by jlmaccal** [comment]
> Sorry, I didn't see the dangerousness of commenting out the timeout context. If I understood correctly, the timeout is a useful debugging mechanism which provide information of problematic follower runs in the REMD and kill it after the default timeout time. In the worst case, the problematic run doesn't get killed, the exchange_states_for_energy_calc function call in the leader.py will still block it and the simulation will keep hanging there. Plus, this mechanism doesn't use in older MELD version. Besides, we validated it against previous MELD runs and the results are reproducible.

It's not really about debugging. In the worst case, half of your cluster sits idle for a week doing nothing because a run is hanging. But you're correct that this doesn't change the results at all.

It would be great if one / both of you could try the tiny change in #74 to see if that fixes things. @alberto99 In your case, I think this will just end up giving a RuntimeError that timeouts have expired, rather than the `StopIteration` issue, but at leas that part will be working as expected. I think the rest has to do with MPI, as you suggest.

**Comment by jlmaccal** [comment]
@alberto99 The traceback you gave is related to changes in how python handles `StopIteration` errors inside generators for python >= 3.5. The fix on our end is trivial, as our context manager that implements the timeout should have only been `yield`ing once. The second one was previously harmless, but became an error after 3.5.

I was able to reproduce and debug all of this with a simple script that doesn't use MELD or MPI at all. So at least that part of the problem should be fixed by #74.

**Comment by frank-stonybrook** [comment]
> It's not really about debugging. In the worst case, half of your cluster sits idle for a week doing nothing because a run is hanging. But you're correct that this doesn't change the results at all.

Thanks for the clarification. I was just worried about the correctness of the simulation results when Al refers to it as dangerous. I will less worried about the GPU sitting idle as I hardly encountered those situations.

>  It would be great if one / both of you could try the tiny change in #74 to see if that fixes things

Sure. I will give it a try and let you know whether it helps.

**Comment by jlmaccal** [comment]
Fixed by #74.

**Comment by frank-stonybrook** [comment]
@jlmaccal The fix #74 is working on our cluster with **python 3.7.6** and **OpenMPI**. Thanks!

**Comment by alberto99** [comment]
Thanks Justin,
Al.
--------------------------------------------------
Alberto Perez 
Assistant Professor
University of Florida, Department of Chemistry
https://perez.chem.ufl.edu
